File: tools/cron/smb/process.py
from solana.rpc.api import Client
import pprint
import base58
import json
import requests
import os
from multiprocessing import Pool

pp = pprint.PrettyPrinter(indent=1)
http_client = Client("https://dawn-divine-feather.solana-mainnet.quiknode.pro/971e7d86ae5ad52d9d48097afaec1f7edde191e7/")

TRANSACTION_URI = 'http://localhost:3000/transactions'
COLLECTION_URI = 'http://localhost:3000/collections/'

SMB = 'J7RagMKwSD5zJSbRQZU56ypHUtux8LRDkUpAPSKH4WPp'
LAMPORTS_PER_SOL = 1000000000

def tx_date(tx):
  return tx['result']['blockTime'] * 1000

def get_price(data):
  number = "".join(reversed([data[i:i+2] for i in range(0, len(data), 2)]))
  return int(number, 16) / LAMPORTS_PER_SOL

def get_account(tx, index):
  try:
    return tx['result']['transaction']['message']['accountKeys'][index]
  except:
    return None

def ix_info(tx):
  try:
    instructions = tx['result']['transaction']['message']['instructions']
    for ix in instructions:
      data = base58.b58decode(ix['data']).hex()
      if data[:16] == '856e4aaf709ff59f' or data[:16] == '5f81edf00831df84' or data[:16] == '95e23468068ee627':
        return data, ix['accounts']
    return None, None
  except Exception:
    return None, None

def tx_type(tx):
  data, accounts = ix_info(tx)
  if not data:
    return
  elif data[:16] == '856e4aaf709ff59f' and data[-2:] == '01':
    return 'list'
  elif data[:16] == '856e4aaf709ff59f' and data[-2:] == '00':
    return 'offer'
  elif data[:16] == '5f81edf00831df84':
    if len(tx['result']['meta']['postTokenBalances']) == 0:
      return 'cancel_offer'
    else:
      return 'cancel'
  elif data[:16] == '95e23468068ee627':
    index = accounts[1]
    if get_account(tx, index) == 'So11111111111111111111111111111111111111112':
      return 'accept_offer'
    else:
      return 'buy'

def get_cancel_tx_info(tx):
  _, accounts = ix_info(tx)
  owner_index = accounts[0]

  payload = {}
  payload['date'] = tx_date(tx)
  payload['type'] = 'cancel'
  payload['owner'] = get_account(tx, owner_index)
  payload['new_owner'] = payload['owner']
  payload['mint'] = tx['result']['meta']['postTokenBalances'][0]['mint']
  return payload

def get_list_tx_price(tx):
  data, _ = ix_info(tx)
  if data[:16] == '856e4aaf709ff59f' and data[-2:] == '01':
    return get_price(data[32:48])

def get_list_tx_info(tx):
  _, accounts = ix_info(tx)
  owner_index = accounts[0]
  mint_index = accounts[1]

  payload = {}
  payload['date'] = tx_date(tx)
  payload['type'] = 'list'
  payload['price'] = get_list_tx_price(tx)
  payload['owner'] = get_account(tx, owner_index)
  payload['new_owner'] = payload['owner']
  payload['mint'] = get_account(tx, mint_index)
  return payload

def get_buy_tx_info(tx):
  _, accounts = ix_info(tx)
  new_owner_index = accounts[0]
  owner_index = accounts[7]
  mint_index = accounts[1]

  payload = {}
  payload['date'] = tx_date(tx)
  payload['type'] = 'buy'
  payload['price'] = (tx['result']['meta']['preBalances'][new_owner_index] - tx['result']['meta']['postBalances'][new_owner_index]) / LAMPORTS_PER_SOL
  payload['owner'] = get_account(tx, owner_index)
  payload['new_owner'] = get_account(tx, new_owner_index)
  payload['mint'] = get_account(tx, mint_index)
  return payload

def get_accept_tx_info(tx):
  _, accounts = ix_info(tx)
  new_owner_index = accounts[7]
  owner_index = accounts[0]
  mint_index = accounts[2]
  escrow_index = accounts[5]

  payload = {}
  payload['date'] = tx_date(tx)
  payload['type'] = 'buy'
  payload['subtype'] = 'accept_offer'
  payload['price'] = (tx['result']['meta']['preBalances'][escrow_index] - tx['result']['meta']['postBalances'][escrow_index]) / LAMPORTS_PER_SOL
  payload['owner'] = get_account(tx, owner_index)
  payload['new_owner'] = get_account(tx, new_owner_index)
  payload['mint'] = get_account(tx, mint_index)
  return payload

def get_transaction_info(tx):
  payload = {}
  type = tx_type(tx)
  if type == 'list':
    payload = get_list_tx_info(tx)
  elif type == 'cancel':
    payload = get_cancel_tx_info(tx)
  elif type == 'buy':
    payload = get_buy_tx_info(tx)
  elif type == 'accept_offer':
    payload = get_accept_tx_info(tx)
  payload['type'] = type
  payload['tx'] = tx['result']['transaction']['signatures'][0]
  payload['marketplace'] = 'smb'
  return payload

def get_payload(sig):
  tx = http_client.get_transaction(sig)

  err = 1
  payload = {'tx': sig, 'marketplace': 'smb'}
  try:
    err = tx['result']['meta']['err']
  except:
    # tx has a problem
    return payload
  if err is not None:
    return payload
  payload = get_transaction_info(tx)
  return payload

def process_tx(sig):
  payload = get_payload(sig)

  if 'mint' in payload:
    symbol = requests.get(COLLECTION_URI + payload['mint']).json()
    if symbol:
      payload['symbol'] = symbol['symbol']

  requests.post(TRANSACTION_URI, json=payload)

def get_latest_transactions(until=None):
  before = None
  sigs = []
  while True:
    res = http_client.get_signatures_for_address(
      account = SMB,
      before = before,
      until = until,
      commitment = 'finalized'
    )
    if len(res['result']) == 0:
      return sigs
    for sig in res['result']:
      sigs.append(sig['signature'])
    before = res['result'][-1]['signature']
    if len(res['result']) < 1000:
      return sigs
    print(len(sigs))

def update_last_sig(tx_hash):
  with open('last_sig.json', 'r') as f:
    sigs = json.load(f)
  sigs.append(tx_hash)
  with open('last_sig.json', 'w') as f:
    json.dump(sigs, f)

if __name__ == '__main__':
  abspath = os.path.abspath(__file__)
  dname = os.path.dirname(abspath)
  os.chdir(dname)

  with open('last_sig.json', 'r') as f:
    sigs = json.load(f)
  until = sigs[-1]

  sigs = get_latest_transactions(until)
  pool = Pool(2)
  pool.map(process_tx, sigs)
  pool.close()
  pool.join()

  # only update last_sig once all data is uploaded
  update_last_sig(sigs[0])

File: cron/solanart/process.ts
// @ts-nocheck
import {LAMPORTS_PER_SOL, PublicKey} from "@solana/web3.js";
import {BN} from "@project-serum/anchor";
import {
  getLatestTxsUntilBlockTime,
  timer,
  uploadTxToDB,
  checkUpgradeAndHalt,
  base58
} from '../../tools/utils'

const SOLANART = new PublicKey('CJsLwbP1iu5DuUikHEJnLfANgKy6stB2uFgvBBHoyxwz')

class Assignable {
    constructor(properties) {
        Object.keys(properties).map((key) => {
            return (this[key] = properties[key]);
        });
    }
}

class tx_info extends Assignable {

}

function getMintFromTx(tokenBalances, index) {
  let mint;
  tokenBalances.forEach((t) => {
    if (t.accountIndex == index) {
      mint = t.mint
    }
  })
  return mint
}

function getAcceptOfferTxInfo(tx, ix, index) {
  let accounts = ix.accounts;
  let res = new tx_info({
      mint: accounts[3].toBase58(),
      owner: accounts[0].toBase58(),
      price: new BN(base58.decode(ix.data).slice(1, 9), 'le').toNumber() / LAMPORTS_PER_SOL,
      date: tx.blockTime * 1000 + index,
      marketplace: 'solanart',
      escrow: accounts[1].toBase58(),
      type: 'buy',
      subtype: 'accept_offer',
      ix: index,
      tx: tx.transaction.signatures[0],
  })
  return res
}

function getListTxInfo(tx, ix, index) {
  let accounts = ix.accounts;
  let res = new tx_info({
    mint: accounts[4].toBase58(),
    owner: accounts[0].toBase58(),
    price: new BN(base58.decode(ix.data).slice(1, 9), 'le').toNumber() / LAMPORTS_PER_SOL,
    date: tx.blockTime * 1000,
    marketplace: 'solanart',
    escrow: accounts[3].toBase58(),
    type: 'list',
    tx: tx.transaction.signatures[0],
    ix: index
  })
  return res
}

function getBuyTxInfo(tx, ix, index) {
    let accountKeys = tx.transaction.message.accountKeys
    let accounts = ix.accounts;
    let mintIdx;
    accountKeys.forEach((k,index)=>{
        k.pubkey.toBase58() == accounts[1].toBase58() ? mintIdx = index : null;
    })

    let type;
    accounts[3].toBase58() == accounts[0].toBase58() ? type = 'cancel' : type = 'buy';
    // let price = new BN(base58.decode(ix.data).slice(1, 9), 'le').toNumber() / LAMPORTS_PER_SOL;
    let taIndex;
    accountKeys.forEach((k, index) => {
      k.pubkey.toBase58() == accounts[1].toBase58() ? taIndex = index : null;
    })
    let price = (tx.meta.preBalances[0] - tx.meta.postBalances[0] - tx.meta.postBalances[taIndex] - tx.meta.fee) / LAMPORTS_PER_SOL;
    if (type == 'cancel') {
      price = 0
    }

    let res = new tx_info({
        mint: getMintFromTx(tx.meta.postTokenBalances, mintIdx),
        owner: accounts[3].toBase58(),
        new_owner: accounts[0].toBase58(),
        price: price,
        date: tx.blockTime * 1000,
        marketplace: 'solanart',
        escrow: accounts[4].toBase58(),
        type: type,
        tx: tx.transaction.signatures[0],
        ix: index
    })
    return res
}

function getUpdateTxInfo(tx, ix, index) {

    let accounts = ix.accounts;

    let res = new tx_info({
        mint: accounts[2].toBase58(),
        owner: accounts[0].toBase58(),
        price: new BN(base58.decode(ix.data).slice(1, 9), 'le').toNumber() / LAMPORTS_PER_SOL,
        date: tx.blockTime * 1000,
        marketplace: 'solanart',
        escrow: accounts[1].toBase58(),
        type: 'list',
        subtype: 'update',
        tx: tx.transaction.signatures[0],
        ix: index
    })
    return res
}

async function processLatestTxs(latestTxs) {
  let payloads = []
  for (let i = 0; i < latestTxs.length; i++) {
    try {
      let tx = latestTxs[i]
      for (let index = 0; index < tx.transaction.message.instructions.length; index++) {
        try {
          let ix = tx.transaction.message.instructions[index]
          await checkUpgradeAndHalt(ix, SOLANART)
          if (ix.programId.toBase58() == SOLANART.toBase58()) {
            let ixData = base58.decode(ix.data)
              let type = new BN(ixData[0]).toNumber()
              switch (type) {
                case 2:
                  payloads.push(getAcceptOfferTxInfo(tx, ix, index))
                  break;
                case 4:
                  payloads.push(getListTxInfo(tx, ix, index))
                  break;
                case 5:
                  payloads.push(getBuyTxInfo(tx, ix, index))
                  break;
                case 6:
                  payloads.push(getUpdateTxInfo(tx, ix, index))
                  break;
                default:
                  break;
            }
          }
        } catch {}
      }
    } catch {}
  }
  return payloads
}

async function magiceden(until) {
  const txs = await getLatestTxsUntilBlockTime(SOLANART, until)
  const payloads = await processLatestTxs(txs)
  for (let i = 0; i < payloads.length; i++) {
    await uploadTxToDB(payloads[i])
  }
  return txs.length > 0 ? txs[txs.length - 1].blockTime : until
}

async function main() {
  let until = 1645483959
  while (true) {
    let prev_until = until
    await timer(1000)
    try {
      until = await magiceden(until)
    } catch (e) {
      until = prev_until
    }
  }
}

console.log('Running client.');

main().then(() => console.log(''));
